exasol-jdbc/src/main/scala/com/exasol/spark/DefaultSource.scala

```
@@ -154,11 +155,37 @@ class DefaultSource
     options: ExasolOptions,
     manager: ExasolConnectionManager
   ): Unit = {
-    val writer = new ExasolWriter(sqlContext.sparkContext, tableName, df.schema, options, manager)
-    val exaNodesCnt = writer.startParallel()
-    val newDF = repartitionPerNode(df, exaNodesCnt)
+    val maxParallel = sqlContext.sparkContext.defaultParallelism
+    logInfo(s"saveDataFrame, maxParallellelism=$maxParallel")
+    val mainConnection = manager.writerMainConnection()
+
+    if (mainConnection == null) {
+      throw new RuntimeException(
+        ExaError
+          .messageBuilder("F-SEC-7")
+          .message("Could not create main JDBC connection to Exasol cluster.")
+          .mitigation("Please make sure that there network connection between Spark and Exasol clusters.")
+          .toString()
+      )
+    }
 
-    newDF.rdd.foreachPartition(iter => writer.insertPartition(iter))
+    try {
+      val exaNodesCnt = manager.initParallel(mainConnection, maxParallel)
+      val hosts = manager.subConnections(mainConnection)
+      val newDF = repartitionPerNode(df, exaNodesCnt)
+      val writer = new ExasolWriter(sqlContext.sparkContext, tableName, df.schema, options, hosts, manager)
+
+      logInfo(s"save with nodes=$exaNodesCnt")
+      newDF.rdd.foreachPartition(iter => writer.insertPartition(iter))
+      mainConnection.commit()
+    } catch {
+      case ex: Exception => {
+        logError("Exception during writing, roll back transaction", ex)
+        mainConnection.rollback()
+      }
```

Should we re-throw the exception here?

---

Great point! Yes, we should